client/client.c

```diff
@@ -0,0 +1,202 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include <unistd.h>
+#include <arpa/inet.h> 
+#include <error.h>
+#include <argp.h>
+
+static const uint16_t MSG_MAX_LEN = (uint16_t)~0U;
+
+/* Argp's global variables. */
+const char *argp_program_version =
+	"Gatekeeper dynamic configuration client 1.0";
+
+/* Arguments. */
+static char adoc[] = "<PATH>";
+
+static char doc[] = "Gatekeeper Client -- configure the Gatekeeper via "
+	"dynamic configuration";
+
+static struct argp_option options[] = {
+	{"server-path",		's',	"FILE",		0,
+		"Path to create the UNIX socket",	1},
+	{ 0 }
+};
+
+struct args {
+	const char *server_path;
+	const char *lua_script_path;
+};
+
+static error_t
+parse_opt(int key, char *arg, struct argp_state *state)
+{
+	struct args *args = state->input;
+
+	switch (key) {
+	case 's':
+		args->server_path = arg;
+		break;
+
+	case ARGP_KEY_INIT:
+		args->lua_script_path = NULL;
+		break;
+
+	case ARGP_KEY_ARG:
+		if (args->lua_script_path) {
+			argp_error(state,
+				"Wrong number of arguments; only one is allowed");
+		}
+		args->lua_script_path = arg;
+		break;
+
+	case ARGP_KEY_END:
+		if (!args->lua_script_path) {
+			argp_error(state,
+				"The lua script path was not specified");
+		}
+		break;
+
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static struct argp argp = {options, parse_opt, adoc, doc, NULL, NULL, NULL};
+
+int
+main(int argc, char *argv[])
+{
+	int ret = 0;
+	int sock_fd;
+	FILE *fp;
+	char *line = NULL;
+	char file_str[MSG_MAX_LEN];
+	char send_buff[MSG_MAX_LEN];
+	char recv_buff[MSG_MAX_LEN];
+	size_t len = 0;
+	size_t read_len;
+	size_t total_file_len = 0;
+	uint16_t msg_len;
+	int send_size = 0;
+	int recv_size = 0;
+	int tot_send = 0;
+	int tot_recv;
+	struct sockaddr_un serv_addr; 
+
+    	struct args args = {
+		/* Defaults. */
+		.server_path = "/tmp/dyn_cfg.socket",
+	};
+
+	/* Read parameters. */
+	argp_parse(&argp, argc, argv, 0, NULL, &args);
+
+	if((sock_fd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
+		perror("Error : Could not create socket\n");
+		return -1;
+	}
+
+	memset(&serv_addr, '0', sizeof(serv_addr)); 
+	serv_addr.sun_family = AF_UNIX;
+	strcpy(serv_addr.sun_path, args.server_path);
+
+	if(connect(sock_fd, (struct sockaddr *)&serv_addr,
+			sizeof(serv_addr)) < 0) {
+		perror("Error : Connect failed\n");
+		ret = -1;
+		goto out;
+	}
+
+	memset(file_str, 0, sizeof(file_str));
+
+	if ((fp = fopen(args.lua_script_path, "r")) == NULL) {
+		printf("Error: Failed to open file %s - %s\n",
+			args.lua_script_path, strerror(errno));
+		ret = -1;
+		goto out;
+	}
+
+	while ((read_len = getline(&line, &len, fp)) != -1) {
+		memcpy(file_str + total_file_len, line, read_len);
+		total_file_len += read_len;
+	}
```

Comment:
1. Read the whole file at once with `fread()`, and make sure it fits the buffer.

2. For readability, put the code that put the file in the buffer in a function.

3. Read the file before you get a connection. There's no reason to keep Gatekeeper waiting.